roofit/roofitcore/inc/RooFormula.h

```diff
@@ -62,31 +66,23 @@ class RooFormula : public ROOT::v5::TFormula, public RooPrintable {
   virtual void printArgs(std::ostream& os) const ;
   void printMultiline(std::ostream& os, Int_t contents, Bool_t verbose=kFALSE, TString indent="") const ;
 
-  inline virtual void Print(Option_t *options= 0) const {
+  virtual void Print(Option_t *options= 0) const {
     // Printing interface (human readable)
     printStream(defaultPrintStream(),defaultPrintContents(options),defaultPrintStyle(options));
   }
 
-protected:
-  
-  RooFormula& operator=(const RooFormula& other) ;
-  void initCopy(const RooFormula& other) ;
-
-  // Interface to ROOT::v5::TFormula engine
-  Int_t DefinedVariable(TString &name, int& action) ; // ROOT 4
-  Int_t DefinedVariable(TString &name) ; // ROOT 3
-  Double_t DefinedValue(Int_t code) ;
+private:
+  RooFormula& operator=(const RooFormula& other);
+  std::string processFormula(std::string origFormula) const;
+  RooArgList  usedVariables() const;
+  std::string reconstructFormula(std::string internalRepr) const;
+  std::vector<bool> findCategoryServers(const RooAbsCollection& collection) const;
 
-  RooArgSet* _nset ;
-  mutable Bool_t    _isOK ;     // Is internal state OK?
-  RooLinkedList     _origList ; //! Original list of dependents
-  std::vector<Bool_t> _useIsCat;//! Is given slot in _useList a category?
-  RooLinkedList _useList ;      //! List of actual dependents 
-  mutable RooArgSet _actual;    //! Set of actual dependents
-  RooLinkedList _labelList ;    //  List of label names for category objects  
-  mutable Bool_t    _compiled ; //  Flag set if formula is compiled
+  RooArgList _origList; //! Original list of dependents
+  std::vector<bool> _isCategory; //! Whether an element of the _origList is a category.
+  std::unique_ptr<TFormula> _tFormula; //! The formula used to compute values
 
-  ClassDef(RooFormula,1)     // ROOT::v5::TFormula derived class interfacing with RooAbsArg objects
+  ClassDef(RooFormula,0)
```

Comment:
0 indicates that there is no intent of 'storing' this kind of objects, is that the case?

Reply:
Yes. The formula instance is constructed on-demand by the owning classes.

Thanks for pointing this out, though, because if this hadn't been the intention, the error would have been hard to spot.